feat: implement comprehensive auth tests and fix failing unit tests

LeidejanedaRosa opened the pull request:
## 📋 Resumo das Alterações

### ✨ Implementação de Testes de Autenticação e JWT

#### 🔐 Testes de Auth Implementados:
- **AuthController tests** (`auth.controller.spec.ts`):
  - Testes de login para usuário e empresa
  - Validação de credenciais inválidas
  - Testes de email não confirmado
  - Validação de formato de dados
  - Testes de campos obrigatórios
  - Testes do endpoint `userLogged`

- **AuthLoginService tests** (`auth-login.service.spec.ts`):
  - Login de usuário com credenciais válidas
  - Login de empresa com credenciais válidas
  - Tratamento de email não confirmado
  - Validação de usuário/empresa inexistente
  - Validação de senha inválida
  - Sanitização de dados sensíveis (remoção de password, tokens)

- **JwtStrategy tests** (`jwt.strategy.spec.ts`):
  - Validação de payload JWT para usuários
  - Validação de payload JWT para empresas
  - Tratamento de usuário/empresa não encontrada
  - Remoção de campos sensíveis das respostas
  - Tratamento de payload inválido

- **Auth E2E tests** (`auth.e2e-spec.ts`):
  - Testes de integração completos do endpoint `/auth/login`
  - Validação de responses HTTP
  - Testes de diferentes cenários de login

#### 🎭 Mocks Criados:
- **company.mock.ts**: Mocks para empresa, empresa não confirmada e entidade empresa
- **user-login.mock.ts**: Mocks para login de usuário, empresa e cenários inválidos
- **jwt-payload.mock.ts**: Mocks para payloads JWT válidos e inválidos

#### 🏗️ Estrutura de Auth Implementada:
- **principal.types.ts**: Tipos TypeScript para usuários e empresas autenticadas
- **principal.mapper.ts**: Mappers para conversão de entidades para principals
- **auth-login.service.ts**: Service completo de autenticação com JWT

### 🔧 Correções de Estrutura e Prettier

#### 📁 Reorganização de Arquivos:
- Renomeou `company-repository.ts` → `company.repository.ts`
- Moveu `jwt.strategy.ts` de `jtw/` → `jwt/` (correção de typo)
- Padronizou nomenclatura de arquivos de repositório

#### 📦 Correções de Importações:
- Corrigiu importações nos módulos: `auth.module.ts`, `company.module.ts`, `jobs.module.ts`, `user.module.ts`
- Atualizou imports nos services: `activate-company.service.ts`, `create-company.service.ts`, etc.
- Corrigiu paths relativos para imports absolutos onde necessário

#### ⚙️ Configurações Atualizadas:
- **package.json**: Configuração completa do Jest com coverage e moduleNameMapper
- **.env.example**: Corrigiu `SECRET_KEY` → `JWT_SECRET`
- **jest-e2e.json**: Adicionou moduleNameMapper para testes e2e

#### 🎨 Formatação Prettier:
- Aplicou formatação consistente em todos os arquivos modificados
- Padronizou indentação e quebras de linha
- Organizou imports alfabeticamente
- Removeu trailing commas inconsistentes

### 🧪 Correção de Testes Falhando

#### 🔧 Testes Corrigidos:
- **savedjobs.service.spec.ts**: Adicionou mocks de repositórios TypeORM
- **candidacy.service.spec.ts**: Implementou CandidacyRepositoryMock
- **create-user.service.spec.ts**: Corrigiu mocks e cenários de teste
- **find-all-users.service.spec.ts**: Atualizou PageOptionsDto
- **recovery-password-by-email.service.spec.ts**: Corrigiu fluxo de recuperação
- **update-password-by-email.service.spec.ts**: Validação de senhas
- **update-user.service.spec.ts**: Testes de upload de arquivo

#### 🗑️ Limpeza:
- Removeu testes obsoletos: `delete-user.service.spec.ts`, `find-one-user.service.spec.ts`
- Atualizou mocks de user para consistência

## 📊 Resultados dos Testes
- ✅ **10 test suites** passando
- ✅ **47 testes** passando (incluindo 25+ novos testes de auth)
- ✅ **0 testes** falhando
- ✅ **Cobertura de testes** mantida acima de 85%
- ✅ **E2E tests** funcionando

## 🔍 Commits Organizados
1. **refactor**: fix file structure, import paths and test config
2. **feat**: implement auth and JWT tests  
3. **fix**: correct failing unit tests

## 🎯 Benefícios
- Cobertura completa de testes para módulo de autenticação
- Maior confiabilidade no sistema de login
- Estrutura de código mais organizada e consistente
- Base sólida para futuras implementações de auth
- Redução de bugs relacionados à autenticação

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - JWT configurável por ambiente (JWT_SECRET) com emissor/audiência/algoritmo; login passa a retornar { token, info }. Endpoint autenticado para excluir "minha empresa" (DELETE /company/me). Novos mapeadores de principal.

- **Changes**
  - Normalização de e-mail em cadastros; duplicatas tratadas com conflitos (409/Conflict). Erros de autenticação e senhas agora lançam exceções HTTP apropriadas. Controladores retornam diretamente respostas com decorators de status.

- **Bug Fixes**
  - Listagem de vagas da empresa logada sempre retorna 200 com array (vazio se necessário).

- **Database**
  - Nova migração adiciona índice único em e-mail de empresas (normalizado); ambiente de testes usa sqlite in-memory.

- **Documentation**
  - Swagger documenta exclusão da própria empresa (DELETE /company/me).

- **Tests**
  - Ampliação de testes E2E/unidade e novos helpers/mocks para cobertura de auth, JWT, usuários, empresas e repositórios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

LeidejanedaRosa committed:
refactor: fix file structure and import paths

LeidejanedaRosa committed:
refactor: fix file structure, import paths and test config

LeidejanedaRosa committed:
feat: implement auth and JWT tests

LeidejanedaRosa committed:
fix: correct failing unit tests

LeidejanedaRosa committed:
fix: replace hardcoded passwords with centralized test constants

LeidejanedaRosa committed:
fix: replace hardcoded IPs and phone numbers with test constants

LeidejanedaRosa committed:
fix: implement secure SSL config and proper array sorting

LeidejanedaRosa committed:
fix: replace hardcoded IPs with RFC-compliant documentation addresses

MikaelMelo1 commented on src/modules/auth/auth.module.ts:
Da uma olhada nesse module, sobre o exports se é necessário ou não. O restante está ok.

LeidejanedaRosa replied:
O "export class AuthModule {}", na linha 34, é necessário porque:

É obrigatório para módulos NestJS, todas as classes de módulo precisam ser exportadas para poderem ser importadas em outros arquivos. E ele é usado no app.module.ts: O AppModule importa o AuthModule com import { AuthModule } from './modules/auth/auth.module'

MikaelMelo1 commented on src/modules/company/repository/company.repository.ts:
Da uma olhadinha na parte do findCompanyById e findOneById, eles fazem a mesma coisa?
Da pra também pra otimizar os métodos de update, reduzindo as chamadas ao banco.

LeidejanedaRosa replied:
Fluxo alterado:
Antes: find → verificar se existe → update/delete → find novamente
Agora: update/delete direto → verificar result.affected → buscar apenas se necessário

Corrigido!

MikaelMelo1 commented on src/modules/company/services/delete-company.service.ts:
Podemos refatorar essa parte, passando assim a responsabilidade pro controller, usando o @httpcode() ? Assim fica mais separado cada responsabilidade

LeidejanedaRosa replied:
Refatorado!

MikaelMelo1 commented on src/modules/company/services/update-company.service.ts:
```diff
@@ -30,7 +30,7 @@ export class UpdateCompanyService {
     delete data.file;
 
-    await this.companyRepository.UpdateCompanyById(company.id, data);
+    await this.companyRepository.updateCompanyById(company.id, data);
```
da uma conferida se é isso mesmo, se foi realmente trocado para caixa baixa a primeira letra.

LeidejanedaRosa replied:
Sim, foi sim! Foi corrigido pelo Eslint, camelCase para métodos e variáveis.

MikaelMelo1 commented on src/modules/jobs/services/get-all-jobs-from-logged-company.service.ts:
Da pra melhorar a arquitetura que nem solicitei do Delete, movendo a responsabilidade pro Controller, assim vai separar mais a responsabilidade. A dependência ComapnyRepository ta sendo utilizada?

LeidejanedaRosa replied:
O CompanyRepository realmente não estava sendo utilizado, foi removido.

Eu fiz:
- Removi CompanyRepository (não está sendo usado)
- Simplifiquei o service para retornar dados diretamente
- Movi responsabilidade do status HTTP para o controller
- Adicionei decorator @HttpCode

MikaelMelo1 commented on src/modules/savedjobs/services/savedjobs.service.spec.ts:
Podemos pensar em deixar de usar o mock compartilhado e usar mocks individuais pra cada repositório.

LeidejanedaRosa replied:
Feito!

MikaelMelo1 commented on src/modules/user/services/create-user.service.ts:
Podemos colocar um ConflictException no lugar do 404, o nest vai converter pra um http

LeidejanedaRosa replied:
Alteração feita!

MikaelMelo1 commented on test/modules/auth/services/auth-login.service.spec.ts:
Cobertura de testes muito bom, podemos mudar a parte do user não existente trocando a mensagem por exemplo pela email ou senha não conferem... assim evitamos um risco de segurança

LeidejanedaRosa replied:
Alteração feita!

MikaelMelo1 commented on test/modules/auth/strategy/jwt.strategy.spec.ts:
Muito bom.
Na verificação da pra trocar o validate por rejects.toThrow() da pra fazer as duas coisas em uma linha só (caso queira implementar, nada de necessidade)

LeidejanedaRosa replied:
Implementado!

MikaelMelo1 commented on test/modules/user/services/create-user.service.spec.ts:
Da pra gente alterar a parte de testes de erro fazendo que eles esperem uma exceção
O uso do jest.spyOn por cima do mock que já tem fica redundante. Da pra remover e usar diretamente (toHaveBeenCalledWith)

LeidejanedaRosa replied:
Implementado!

MikaelMelo1 commented on test/modules/user/services/update-password-by-email.service.spec.ts:
Podemos lançar algumas exceções como notfoundexce.... badrequest.... e também a mesma função do spyOn. (caso necessite)

LeidejanedaRosa replied:
Implementado!

MikaelMelo1 commented on test/modules/auth/controller/auth.controller.spec.ts:
Da pra gente fazer que com o AuthLoginService  lance algumas exceções caso de erro  e retorne os dados  caso de sucesso (token, info)

LeidejanedaRosa replied:
Exceções implementadas.

MikaelMelo1 commented on test/modules/auth/auth.e2e-spec.ts:
Podemos tentar deixar o UserRepo e o Compa..Repo... sem os mocks pq já estamos mockando com o AuthLoginService...
Podemos também lançar exceções no AuthLoginService.

LeidejanedaRosa replied:
Implementei as exceções e removi os mocks desnecessários.

MikaelMelo1 commented on test/mocks/user/get-all-user.mock.ts:
Podemos da uma olhada na parte dos IDs para que não contenha IDs duplicados. Restante ok.

LeidejanedaRosa replied:
Implementado!

MikaelMelo1 commented on test/mocks/auth/company.mock.ts:
Podemos analisar o companymock e companyentitymock eles são quase iguais, podemos deixar o companyentitymock pro registro completo (com senha) e uma outra publiccompanymock pros dados publi... sem senha.

LeidejanedaRosa replied:
Alteração feita!

MikaelMelo1 commented on test/config/test-constants.ts:
o UUID do USER_ID e Company_ID estão com o msm valor podemos trocar.
Podemos renomear o USER e VALID_USER

LeidejanedaRosa replied:
Alteração feita!

MikaelMelo1 commented on test/app.e2e-spec.ts:
Podemos analisar também se é necessário alguns mocks na parte do TestingModule tirando os mocks do mailservice e userrepo...

LeidejanedaRosa replied:
Mocks desnecessários removidos.

MikaelMelo1 reviewed (changes requested):
Alguns pontos para corrigir.

LeidejanedaRosa committed:
perf(company): optimize database queries in the repository

LeidejanedaRosa committed:
refactor(company): simplify delete service using @HttpCode decorator

LeidejanedaRosa committed:
test: refactor auth-login tests to use individual mocks instead of shared mock classes

LeidejanedaRosa committed:
refactor(company): update controller to use @HttpCode decorator in delete endpoint

LeidejanedaRosa committed:
refactor(test): modernize test constants with unique UUIDs and standardized naming

- Replace duplicate UUIDs with unique identifiers for USER_ID and COMPANY_ID
- Rename TEST_EMAILS.USER to TEST_EMAILS.DEFAULT_USER for clarity
- Maintain RFC-compliant IP addresses for documentation
- Improve test isolation and prevent ID conflicts

LeidejanedaRosa committed:
refactor(test): modernize mock structure and organization

LeidejanedaRosa committed:
test(auth): modernize authentication test suite with factory pattern

LeidejanedaRosa committed:
test(user): modernize user service tests with factory pattern

LeidejanedaRosa committed:
test(e2e): simplify and modernize end-to-end tests

LeidejanedaRosa committed:
test(modules): modernize candidacy and savedjobs service tests

LeidejanedaRosa committed:
refactor(services): improve error handling and security in core services

LeidejanedaRosa committed:
feat(controllers): add @HttpCode decorators and improve API consistency

LeidejanedaRosa committed:
refactor(modules): optimize jobs module dependencies and structure

LeidejanedaRosa committed:
feat(auth): enhance JWT security with structured payload

LeidejanedaRosa committed:
feat(database): add PostgreSQL environment validation and port validation

LeidejanedaRosa committed:
fix(company): restore consistent response pattern for DELETE endpoint

LeidejanedaRosa committed:
fix(auth): restore @LoggedAdmin authorization context in user controller

LeidejanedaRosa committed:
fix(jobs): restore interface contract and consistent response pattern

LeidejanedaRosa committed:
test: improve TypeScript typing and test patterns

LeidejanedaRosa committed:
feat(company): enhance delete company service with proper error handling

LeidejanedaRosa committed:
refactor(database): improve port validation with better type checking

LeidejanedaRosa committed:
fix(security): remove hard-coded JWT secret in tests

LeidejanedaRosa committed:
feat(database): add unique constraint for companies email

LeidejanedaRosa committed:
feat(security): enhance JWT strategy validation

LeidejanedaRosa committed:
feat(services): improve user and company services

LeidejanedaRosa committed:
refactor(tests): consolidate mock infrastructure and eliminate duplication

LeidejanedaRosa committed:
refactor(database): improve database configuration and entities

LeidejanedaRosa committed:
chore(project): improve application structure and health check

LeidejanedaRosa committed:
fix(auth): substituir tokens hardcoded por constantes de teste

LeidejanedaRosa committed:
refactor(test): centralize password hashing in test constants

LeidejanedaRosa committed:
refactor(swagger): update company delete endpoint documentation

LeidejanedaRosa committed:
security(auth): add type guards and payload normalization in JWT strategy

LeidejanedaRosa committed:
feat(migration): improve email uniqueness with functional index

LeidejanedaRosa committed:
fix(database): resolve duplicate unique constraints issue

LeidejanedaRosa committed:
fix: make companies email unique index partial and non-transactional
